File: python/day6/solution.py
from python.read import parse_lines


def solution(data: str, marker_len: int) -> int:
    marker = []
    index = 0

    while len(marker) < marker_len:
        letters = data[index : marker_len + index]
        if len(set(letters)) == marker_len:
            marker = letters
        index += 1

    return index + marker_len - 1


def solution1(data: list[str]) -> int:
    return solution(data[0], 4)


def solution2(data: list[str]) -> int:
    return solution(data[0], 14)


if __name__ == "__main__":
    print("Day6")
    print(
        f"Marker starts at char no. {solution1(parse_lines('../../inputs/day6.txt'))}"
    )
    print(
        f"Message starts at char no. {solution2(parse_lines('../../inputs/day6.txt'))}"
    )

File: rust/src/day6/solution.rs
use std::collections::HashSet;

fn solution(data: String, marker_len: u32) -> u32 {
    let mut marker: Vec<char> = vec![];
    let mut index = 0;

    while marker.len() < marker_len as usize {
        let letters: &str = &data[index as usize..marker_len as usize + index as usize];
        let hash_marker: HashSet<char> = letters.chars().collect();
        if hash_marker.len() == marker_len as usize {
            marker = letters.chars().collect()
        }
        index += 1
    }

    index + marker_len - 1
}

pub fn solution1(data: Vec<String>) -> u32 {
    solution(data[0].clone(), 4)
}

pub fn solution2(data: Vec<String>) -> u32 {
    solution(data[0].clone(), 14)
}

#[test]
fn test_solution1() {
    assert_eq!(solution1(vec!["bvwbjplbgvbhsrlpgdmjqwftvncz".to_string()]), 5);
    assert_eq!(solution1(vec!["nppdvjthqldpwncqszvftbrmjlhg".to_string()]), 6);
    assert_eq!(solution1(vec!["nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg".to_string()]), 10);
    assert_eq!(solution1(vec!["zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw".to_string()]), 11);
}
#[test]
fn test_solution2() {
    assert_eq!(solution2(vec!["bvwbjplbgvbhsrlpgdmjqwftvncz".to_string()]), 23);
    assert_eq!(solution2(vec!["nppdvjthqldpwncqszvftbrmjlhg".to_string()]), 23);
    assert_eq!(solution2(vec!["nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg".to_string()]), 29);
    assert_eq!(solution2(vec!["zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw".to_string()]), 26);
}
